pkg/storagepool/v1alpha1/kubernetes.go

```diff
@@ -0,0 +1,227 @@
+			if err != nil {
+				return nil, err
+			}
+			return clientset.NewForConfig(config)
+		}
+	}
+
+	if k.list == nil {
+		k.list = func(cs *clientset.Clientset,
+			opts metav1.ListOptions) (*apis.StoragePoolList, error) {
+			return cs.OpenebsV1alpha1().
+				StoragePools().
+				List(opts)
+		}
+	}
+
+	if k.get == nil {
+		k.get = func(cs *clientset.Clientset,
+			name string, opts metav1.GetOptions) (*apis.StoragePool, error) {
+			return cs.OpenebsV1alpha1().
+				StoragePools().
+				Get(name, opts)
+		}
+	}
+
+	if k.create == nil {
+		k.create = func(cs *clientset.Clientset,
+			storagePoolObj *apis.StoragePool) (*apis.StoragePool, error) {
+			return cs.OpenebsV1alpha1().
+				StoragePools().
+				Create(storagePoolObj)
+		}
+	}
+
+	if k.patch == nil {
+		k.patch = func(cs *clientset.Clientset, name string,
+			pt types.PatchType, patchObj []byte) (*apis.StoragePool, error) {
+			return cs.OpenebsV1alpha1().
+				StoragePools().
+				Patch(name, pt, patchObj)
+		}
+	}
+
+	if k.del == nil {
+		k.del = func(cs *clientset.Clientset, name string,
+			opts *metav1.DeleteOptions) error {
+			return cs.OpenebsV1alpha1().
+				StoragePools().
+				Delete(name, opts)
+		}
+	}
+}
+
+// WithClientset sets the kubernetes clientset against
+// the kubeclient instance
+func WithClientset(c *clientset.Clientset) KubeclientBuildOption {
+	return func(k *Kubeclient) {
+		k.clientset = c
+	}
+}
+
+// NewKubeClient returns a new instance of kubeclient meant for
+// StoragePool related operations
+func NewKubeClient(opts ...KubeclientBuildOption) *Kubeclient {
+	k := &Kubeclient{}
+	for _, o := range opts {
+		o(k)
+	}
+	k.withDefaults()
+	return k
+}
+
+// getClientsetOrCached returns either a new instance
+// of kubernetes client or its cached copy
+func (k *Kubeclient) getClientsetOrCached() (*clientset.Clientset, error) {
+	if k.clientset != nil {
+		return k.clientset, nil
+	}
+	c, err := k.getClientset()
+	if err != nil {
+		return nil,
+			errors.WithStack(errors.Wrap(err, "failed to get clientset"))
+	}
+	k.clientset = c
+	return k.clientset, nil
+}
+
+// List returns a list of StoragePool
+// instances present in kubernetes cluster
+func (k *Kubeclient) List(opts metav1.ListOptions) (*apis.StoragePoolList, error) {
+	cs, err := k.getClientsetOrCached()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list storage pool")
+	}
+	return k.list(cs, opts)
+}
+
+// Get returns an StoragePool instance from kubernetes cluster
+func (k *Kubeclient) Get(name string, opts metav1.GetOptions) (*apis.StoragePool, error) {
+	if strings.TrimSpace(name) == "" {
+		return nil, errors.New("failed to get storage pool: missing name")
+	}
+	cs, err := k.getClientsetOrCached()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get storage pool: {%s}", name)
+	}
+	return k.get(cs, name, opts)
+}
+
+// Create creates an StoragePool instance in kubernetes cluster
+func (k *Kubeclient) Create(obj *apis.StoragePool) (*apis.StoragePool, error) {
```

Comment:
Is it necessary to check `nil` over `SP` object.

Reply:
client library already has a check for it. For better error handling we can add it.
UT are there for this condition. IMO better to add it.
added.